### ISSUE TITLE
Add 'generic' flag for signal connections

### ISSUE THREAD

**Issue by kylekyleton** [godotengine/godot-proposals#10300]
### Describe the project you are working on

A 3D action adventure game.

### Describe the problem or limitation you are having in your project

A scene is waiting for some events, defined by an external resource:

1. Player has completed an attack combo of more than 5 strikes.
2. Enemy_1 has taken more than 20 units of damage.
3. Enemy_2 has died.
4. Player has jumped.

When all events have taken place, with their parameters met, the scene should call a function and do something cool. These events are just examples, they could be anything. Signals seem like the natural tool for the scene to listen for these events:

- Player has signal `combo_finished(number_of_hits, total_damage)`
- Enemies have signals `damaged(amount)` and `died`
- Player has signal `jumped`

However, given that there could be many, many potential events to listen for, using signals in this way is very difficult. There isn't a good way for a node to connect to and listen for arbitrary signals with arbitrary argument counts and signatures.
Writing a discrete method for each signal isn't a good solution.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow methods to connect to signals generically. That is, it connects to a method and passes three things as arguments:

1. A reference to the node that's emitting the signal
2. The signal's name
3. The signal's arguments, if any, in an array

This permits one method to handle all signals in a way that doesn't require foreknowledge of the signal's arguments.
It means that the method can compare the received signal and its arguments to the list of events its listening for. If the arguments are irrelevant, it's free to ignore them. If the arguments are relevant, it's free to use them.

Using `bind` and `unbind` to reach the right argument count for the connected method requires knowing beforehand how many arguments a signal is going to emit, and how many are relevant -- in the scenario above, it's impossible to know which arguments, if any, will be relevant.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Perhaps something like this:
`Node.signal_name.connect(generic_signal_handler, ConnectFlags.CONNECT_GENERIC)`

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You can emit your signals using `emit_signal.callv([self, &"signal_name", [args]])`
Across many scripts, this would be cumbersome and fragile.

### Is there a reason why this should be core and not an add-on in the asset library?

Signals are a core feature.

**Comment by AThousandShips** [comment]
> Using bind and unbind to reach the right argument count for the connected method requires knowing beforehand how many arguments a signal is going to emit, and how many are relevant

This sounds like basic programming, and pretty normal to know and manage to me

**Comment by kylekyleton** [comment]
> > Using bind and unbind to reach the right argument count for the connected method requires knowing beforehand how many arguments a signal is going to emit, and how many are relevant
> 
> This sounds like basic programming, and pretty normal to know and manage to me

Sure, at connection time, you could read the resource and notice, "Okay, I only need one argument for this connection." And then you can manually look into the script with the relevant Signal (since you can't query the Signal for its arguments beforehand; and if you ever change the Signal's argument count, you also have to manually change your connection parameters) and take note of how many arguments it's gonna send. And then you unbind down to only one -- except `unbind` doesn't allow you to specify _which_ arguments to unbind. What happens if you need to unbind down to one argument, but `unbind` removes the one you care about?

**Comment by AThousandShips** [comment]
> What happens if you need to unbind down to one argument, but unbind removes the one you care about?

How would that be solved by your proposal though? It doesn't explain how it'd know what arguments to ignore and not

But having a "fire and forget" approach to calling methods is a sure way to create issues for yourself, to forget something somewhere because you don't set it up carefully

**Comment by AThousandShips** [comment]
But this would be entirely covered I'd say by vararg functions in GDScript:
* https://github.com/godotengine/godot-proposals/issues/1034

**Comment by kylekyleton** [comment]
> > What happens if you need to unbind down to one argument, but unbind removes the one you care about?
> 
> How would that be solved by your proposal though? It doesn't explain how it'd know what arguments to ignore and not
> 
> But having a "fire and forget" approach to calling methods is a sure way to create issues for yourself, to forget something somewhere because you don't set it up carefully

In my example, the callback method will either use or not use the arguments, according to what the scene has been told to look for in the external Resource. If the generic receiver method receives a signal with the arguments:
`Player, &"combo_finished", [8, 211.3]`, 
it can see, "Oh, that signal from that node is in my list of things to pay attention to. Lemme look through its arguments array and see if anything matches any parameters I might be looking for."
In the example, the scene only cared about the number of hits of the combo, so it can find that information in the signal arguments, passed by array, and it can ignore whatever it doesn't care about.

The point is that whether there are thirty arguments or zero, the callback method will be called successfully, and you can do what you wish with whatever arguments appear in the array, whether that's ignoring them or not.

**Comment by AThousandShips** [comment]
That would have both poor performance and a lot of IMO unnecessary complexity just to avoid planning ahead

**Comment by kylekyleton** [comment]
Yes, I think you are  missing something. It's precisely because I am trying to plan ahead that I'm encountering this limitation.

It's not that _right now_ I have many wildly different connections to make. It's that there are _potentially_ very many. How should a scene be notified if:

- The player has achieved a combo of a certain length
- An actor has taken a certain amount of damage
- An actor has died
- A certain actor has jumped
- A node has been freed
- A certain amount of time has passed
- Two actors have gotten a certain distance apart from each other
- A new actor has entered the scene
- The player used an item, or two, or three
- etc., etc.

...Surely, it's by signals, yes?

In the example I gave in the opening post, where a scene is listening for a series of events like these, how should one expect the scene to be notified of these events taking place? Are signals not the right tool?
If they are the right tool, then the current solution is to write a bespoke method for each signal connection, since each signal has its own arguments that come along with it, and _each of these methods will be doing the same thing:_

Check what signal it is, where it came from, and if its parameters match what's defined in the external Resource.

I realize that I'm arguing from a selfish position. I have this problem and this appears to be a way out. I'm confident that I can figure out another way to solve my problem, and I suppose I will. However, it doesn't seem unreasonable to me to suggest that a Signal be, well, a signal -- that is, that it **signals** to a connected node that something happened: 
>Here's what happened, here's who it happened to, and here's some additional information about the event. Do what you will with that information.

**Comment by KoBeWi** [comment]
`await` ignores argument count, so you can take advantage of that, e.g. like this:
```GDScript
class Awaiter:
	signal finished
	
	func _init(signl: Signal) -> void:
		while true:
			await signl
			finished.emit()
```
Then connect to `awaiter.finished` instead of the target signal. You can wrap it in a nice method that takes Signal and Callable and use it instead of `connect()`. The only problem is that you can't connect it from the editor (at least not easily).

**Comment by passivestar** [comment]
> When all events have taken place, with their parameters met, the scene should call a function and do something cool.

If you need to check that more than 20 damage was done you'll still need code to check for that on the receiving end before 'something cool' happens, right? You can't connect the signal directly to `do_something_cool()`

So you'd have something like a `SomethingCoolManager` node. If your proposal was to be implemented you'd have a generic signal handler like `SomethingCoolManager.something_cool_checker()` with a series of `if/else` branches to check if the received signal has parameters that satisfy your conditions, and only then you'd call `do_something_cool()` from that

But what's stopping you from having several receiving methods on the `SomethingCoolManager`? Why do you want a series of `if/else`s instead of several methods? Just have `on_player_damaged()`, `on_combo_finished()` on it, it will be the same number of lines and about the same amount of coupling

> The only problem is that you can't connect it from the editor (at least not easily)

I think the editor UI for signal connections could be improved, it could at least have an "Unbind All" button that would automatically set the argument count
